include/envoy/filesystem/filesystem.h

```diff
@@ -1,24 +1,106 @@
 #pragma once
 
-#include <cstdint>
-#include <functional>
 #include <memory>
 #include <string>
 
+#include "envoy/common/platform.h"
 #include "envoy/common/pure.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/thread/thread.h"
 
 #include "absl/strings/string_view.h"
 
 namespace Envoy {
 namespace Filesystem {
 
 /**
- * Abstraction for a file on disk.
+ * Abstraction for a basic file on disk.
  */
 class File {
 public:
   virtual ~File() {}
 
+  /**
+   * Open the file with O_RDWR | O_APPEND | O_CREAT
+   * The file will be closed when this object is destructed
+   */
+  virtual void open() PURE;
+
+  /**
+   * Write len bytes from buffer to the file
+   */
+  virtual Api::SysCallSizeResult write(const void* buffer, size_t len) PURE;
+
+  /**
+   * Close the file.
+   */
+  virtual void close() PURE;
+
+  /**
+   * @return bool whether the file is open
+   */
+  virtual bool isOpen() PURE;
+};
+
+typedef std::unique_ptr<File> FilePtr;
+
+/**
+ * Wrapper for platform specific filesystem methods
+ */
+class Instance {
+public:
+  virtual ~Instance() {}
+
+  /**
+   * @return bool whether a file exists on disk and can be opened for read.
+   */
+  virtual bool fileExists(const std::string& path) PURE;
+
+  /**
+   * @return bool whether a directory exists on disk and can be opened for read.
+   */
+  virtual bool directoryExists(const std::string& path) PURE;
+
+  /**
+   * @return ssize_t the size in bytes of the specified file, or -1 if the file size
+   *                 cannot be determined for any reason, including without limitation
+   *                 the non-existence of the file.
+   */
+  virtual ssize_t fileSize(const std::string& path) PURE;
+
+  /**
+   * @return full file content as a string.
+   * @throw EnvoyException if the file cannot be read.
```

Comment:
do we need to throw here? https://github.com/envoyproxy/envoy/blob/master/STYLE.md says:

> Apply caution when using exceptions on the data path for general purpose error handling. Exceptions are not caught on the data path and they should not be used for simple error handling, e.g. with shallow call stacks, where explicit error handling provides a more readable and easier to reason about implementation.

Reply:
Throwing on error here is the existing behavior, we did not want to change that if possible